walnuts_cpp/include/walnuts/nuts.hpp

```diff
@@ -98,19 +102,23 @@ void leapfrog(const F &logp_grad_fun, const Vec<S> &inv_mass, S step,
   logp_next += logp_momentum(rho_next, inv_mass);
 }
 
-template <typename S>
-bool uturn(const Span<S> &span_bk, const Span<S> &span_fw,
-           const Vec<S> &inv_mass) {
+template <Direction D, typename S>
+inline bool uturn(const Span<S> &span_1, const Span<S> &span_2,
+                  const Vec<S> &inv_mass) {
+  auto &&span_bk = (D == Direction::Forward) ? span_1 : span_2;
```

Comment:
Ternary operators don't resolve at compile time even if their conditions could be `constexpr`.   So I would think this would all be more efficient unfolded into two static cases, one forward and one backward.  This could be done with two partial specializations of the template function or a conditional inside.

https://stackoverflow.com/questions/41011900/equivalent-ternary-operator-for-constexpr-if

They can then just delegate to a function by constant reference that assumes one ordering as was being done before.

Reply:
This was just copying the code from `walnuts.hpp` that changed in https://github.com/bob-carpenter/walnuts/pull/18 back into `nuts.hpp`. I'm happy to change it to whatever you'd like, though my personal favorite version is probably just what it was before #18